### DISCUSSION TITLE
How can I configure Data Movement in EKS?

### DISCUSSION THREAD

**Discussion by ryu022304** [vmware-tanzu/velero#7508] [Community support Q&A] [answered]
I installed Velero using the following file and command.
```
$ cat values.yaml
configuration:
  uploaderType: kopia
  backupStorageLocation:
  - name: default
    bucket: <BUCKET_NAME>
    defaultVolumesToFsBackup: False
    config:
      region: ap-northeast-1
    provider: aws
  volumeSnapshotLocation:
  - name: default
    provider: aws
    config:
      region: ap-northeast-1
credentials:
  useSecret: false
features: EnableCSI
defaultSnapshotMoveData: true
defaultVolumeSnapshotLocations: "velero.io/aws:velero-aws"
deployNodeAgent: true
initContainers:
- name: velero-plugin-for-csi
  image: velero/velero-plugin-for-csi:v0.6.3
  imagePullPolicy: IfNotPresent
  volumeMounts:
  - mountPath: /target
    name: plugins
- name: velero-plugin-for-aws
  image: velero/velero-plugin-for-aws:v1.8.2
  volumeMounts:
  - mountPath: /target
    name: plugins
serviceAccount:
  server:
    annotations:
      eks.amazonaws.com/role-arn: "arn:aws:iam::<ACCOUNT>:role/eks-velero-backup"
```
```
$ helm install velero vmware-tanzu/velero  --namespace velero --create-namespace -f values.yaml --version 5.2.2
```

And executing the backup command provided, the STATUS shows as Completed; however, there are warning messages appearing in the logs.

```
$ velero create backup test2 --snapshot-move-data
```
```
$ velero get backup
NAME          STATUS      ERRORS   WARNINGS   CREATED                         EXPIRES   STORAGE LOCATION   SELECTOR
test2         Completed   0        2          2024-03-08 08:52:18 +0900 JST   29d       default            <none>
```
```
$ k -n velero logs velero-6f7bc9f5d6-sbj2q | grep warn
Defaulted container "velero" out of: velero, velero-plugin-for-csi (init), velero-plugin-for-aws (init)
time="2024-03-07T23:52:21Z" level=warning msg="VolumeSnapshotter plugin doesn't support data movement." backup=velero/test2 logSource="pkg/backup/item_backupper.go:528" name=pvc-0882cd24-1317-4bfe-b8f6-ba6a4d7d3c94 namespace= persistentVolume=pvc-0882cd24-1317-4bfe-b8f6-ba6a4d7d3c94 resource=persistentvolumes
time="2024-03-07T23:52:21Z" level=warning msg="VolumeSnapshotter plugin doesn't support data movement." backup=velero/test2 logSource="pkg/backup/item_backupper.go:528" name=pvc-b4521b71-afb9-4ffc-88ad-4aee10fd2c39 namespace= persistentVolume=pvc-b4521b71-afb9-4ffc-88ad-4aee10fd2c39 resource=persistentvolumes
```

How can I configure Data Movement to be executed? Please let me know if any additional information is required."

**Comment by sseago** [comment, marked as answer]
Data movement is only supported for CSI volumes. Since it looks like you have enabled CSI and included the CSI plugin, my guess i that your volumes are not managed by a CSI storage class.

**Reply by ryu022304** [reply]
Thank you for response.
I was able to configure data movement and backup the volumes to s3 by kopia.

**Reply by Jedi-5** [reply]
Hi @ryu022304 @sseago, I have also done above configuration; still following issue occuring:
- /VolumeSnapshotter plugin doesn't support data movement
- /No volume ID returned by volume snapshotter for persistent volume

Please help on above issue, I've disabled CSI and the plugin as well

**Reply by blackpiglet** [reply]
The Velero Native snapshot doesn't support the snapshot data mover.
By far, only the CSI snapshot can support moving the snapshot data.

**Reply by Jedi-5** [reply]
Hi @blackpiglet, Thanks for quick help

we've disabled csi snapshot still following error is coming:
error to connect to backup repo: error to connect repo with storage: error to connect to repository: unable to write config file: unable to create config directory: mkdir /udmrepo: permission denied'

**Reply by blackpiglet** [reply]
Did you use the Kopia as the filesystem uploader to back up the volume data?

The error means the Velero uploader doesn't have the permission to create the backup repository in the OSS.

**Reply by Jedi-5** [reply]
yes, I have used Kopia as fs uploader.
I've configured iam role for service account as per velero's requirement.
could you please let me know about other permission required, which would be grateful

**Reply by blackpiglet** [reply]
https://github.com/vmware-tanzu/velero-plugin-for-aws?tab=readme-ov-file#set-permissions-for-velero
This is the permission required for the AWS environment.

You can find a similar section for GCP and Azure in the following the GitHub repository:
https://github.com/vmware-tanzu/velero-plugin-for-gcp 
https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure

**Reply by Jedi-5** [reply]
I've installed Velero Helm Chart using terraform on EKS, used following module for IRSA:
![irsa](https://github.com/vmware-tanzu/velero/assets/8300079/a8771d34-b48d-4e89-aae0-ae87337a2647)

is it sufficient for the permission?, as it covers the policy required

**Reply by blackpiglet** [reply]
IMO, it's the IRSA-related IAM role's permission that takes effect here.
Please make sure that role's permission is enough for Velero to work.

**Reply by Jedi-5** [reply]
Yes @blackpiglet , mentioned module is having all the permissions required.

**Reply by blackpiglet** [reply]
https://velero.io/docs/main/file-system-backup/#support-readonlyrootfilesystem-setting

Is this related to your scenario?

**Reply by Jedi-5** [reply]
Yes that was the issue, after privileged=true & runAsUser=0; everything working fine.
Thank you very much! @blackpiglet

**Reply by Jedi-5** [reply]
Another issue currently facing:
we have multiple pods with same pv, backup fails in 1st attempt; but success on 2nd attempt

is it due to parallel execution?
could you please let me know how to add timeout or, execution in queue?
